### PR TITLE
[dashboard] Admin dashboard

### PR DESCRIPTION
**Pull request by svenefftinge**
Adds the admin dashboard to the new dashboard.

How to test:
 - create a user with 'admin' flag (default in dev staging)
 - you should see the admin menu on top.
 - try search for users or workspaces.

### CODE REVIEW

File: components/licensor/typescript/ee/src/index.ts [hunk 1]
```diff
@@ -42,23 +42,26 @@ export class LicenseEvaluator {
     }
 
     public validate(): { msg?: string, valid: boolean } {
-        const v = validate(this.instanceID);
-        if (v.valid) {
+        // const v = validate(this.instanceID);
```
Comment (svenefftinge) [inline components/licensor/typescript/ee/src/index.ts:45]: TODO: revert these changes

File: components/dashboard/src/admin/UserSearch.tsx [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+            <div className="flex justify-between w-full">
+                <div className="flex">
+                    <div className="py-4">
+                        <svg className={searching ? 'animate-spin' : ''} width="16" height="16" fill="none" xmlns="http://www.w3.org/2000/svg">
+                            <path fillRule="evenodd" clipRule="evenodd" d="M6 2a4 4 0 100 8 4 4 0 000-8zM0 6a6 6 0 1110.89 3.477l4.817 4.816a1 1 0 01-1.414 1.414l-4.816-4.816A6 6 0 010 6z" fill="#A8A29E" />
+                        </svg>
+                    </div>
+                    <input className="border-0" type="text" placeholder="Search Users" onKeyDown={(ke) => ke.key === 'Enter' && search() } onChange={(v) => { setSearchTerm(v.target.value) }} />
+                </div>
+                <button className="" disabled={searching} onClick={search}>Start Search</button>
```
Comment (gtsiolis) [inline components/dashboard/src/admin/UserSearch.tsx:76]: **suggestion**: No need to prefix this with _Start_. 

```suggestion
                <button className="" disabled={searching} onClick={search}>Search</button>
```

File: components/dashboard/src/admin/WorkspacesSearch.tsx [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            <div className="flex justify-between w-full">
+                <div className="flex">
+                    <div className="py-4">
+                        <svg className={searching ? 'animate-spin' : ''} width="16" height="16" fill="none" xmlns="http://www.w3.org/2000/svg">
+                            <path fillRule="evenodd" clipRule="evenodd" d="M6 2a4 4 0 100 8 4 4 0 000-8zM0 6a6 6 0 1110.89 3.477l4.817 4.816a1 1 0 01-1.414 1.414l-4.816-4.816A6 6 0 010 6z" fill="#A8A29E" />
+                        </svg>
+                    </div>
+                    <input className="border-0" type="text" placeholder="Search in context urls" onKeyDown={(ke) => ke.key === 'Enter' && search() } onChange={(v) => { setSearchTerm(v.target.value) }} />
+                </div>
+                <button className="" disabled={searching} onClick={search}>Start Search</button>
```
Comment (gtsiolis) [inline components/dashboard/src/admin/WorkspacesSearch.tsx:94]: **suggestion**: No need to prefix this with _Start_. 

```suggestion
                <button className="" disabled={searching} onClick={search}>Search</button>
```

File: components/dashboard/src/admin/WorkspacesSearch.tsx [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            </div>
+        </div>
+        <div className="flex flex-col space-y-2">
+            <div className="px-6 py-3 flex justify-between space-x-2 text-sm text-gray-400 border-t border-b border-gray-200">
```
Comment (gtsiolis) [inline components/dashboard/src/admin/WorkspacesSearch.tsx:98]: **nitpick**: Minor margin missing.

```suggestion
            <div className="px-6 py-3 flex justify-between space-x-2 text-sm text-gray-400 border-t border-b border-gray-200 mb-2">
```

File: components/dashboard/src/admin/UserSearch.tsx [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            </div>
+        </div>
+        <div className="flex flex-col space-y-2">
+            <div className="px-6 py-3 flex justify-between space-x-2 text-sm text-gray-400 border-t border-b border-gray-200">
```
Comment (gtsiolis) [inline components/dashboard/src/admin/UserSearch.tsx:80]: **nitpick**: Minor margin missing.

```suggestion
            <div className="px-6 py-3 flex justify-between space-x-2 text-sm text-gray-400 border-t border-b border-gray-200 mb-2">
```

File: components/dashboard/src/admin/UserSearch.tsx [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+function UserEntry(p: { user: User }) {
+    return <Link to={'/admin/users/' + p.user.id}>
+        <div className="rounded-xl whitespace-nowrap flex space-x-2 py-6 px-6 w-full justify-between hover:bg-gray-100 focus:bg-gitpod-kumquat-light group">
+            <div className="pr-3 self-center w-1/12">
+                <img className="rounded-full" src={p.user.avatarUrl} />
+            </div>
+            <div className="flex flex-col w-6/12">
+                <div className="font-medium text-gray-800 truncate hover:text-blue-600">{p.user.fullName}</div>
+                <div className="text-sm overflow-ellipsis truncate text-gray-400 hover:text-blue-600">{User.getPrimaryEmail(p.user)}</div>
```
Comment (gtsiolis) [inline components/dashboard/src/admin/UserSearch.tsx:98]: **suggestion**: This one should be fine without any hover state, WDYT?

```suggestion
                <div className="text-sm overflow-ellipsis truncate text-gray-400">{User.getPrimaryEmail(p.user)}</div>
```

File: components/dashboard/src/admin/UserDetail.tsx [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+
+    return <>
+        <PageWithSubMenu subMenu={adminMenu} title="Users" subtitle="Search and manage all users.">
+            <div className="flex">
+                <div className="flex-1">
+                    <div className="flex"><h3>{user.fullName}</h3>{user.blocked ? <Label text='Blocked' color="red" /> : null} {user.markedDeleted ? <Label text='Deleted' color="red" /> : null}</div>
+                    <p>{user.identities.map(i => i.primaryEmail).filter(e => !!e).join(', ')}</p>
+                </div>
+                <button className="secondary danger ml-3" disabled={activity} onClick={toggleBlockUser}>{user.blocked ? 'Unblock' : 'Block'} User</button>
+                <button className="danger ml-3" disabled={activity} onClick={deleteUser}>Delete User</button>
```
Comment (gtsiolis) [inline components/dashboard/src/admin/UserDetail.tsx:84]: **question**: What do you think of reverting the actions here so that _Block User_ is the primary action instead of _Delete User_. Feel free to leave this as is.

File: components/dashboard/src/admin/UserSearch.tsx [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        } finally {
+            setSearching(false);
+        }
+    }
+    return <PageWithSubMenu subMenu={adminMenu} title="Users" subtitle="Search an manage all users.">
+        <div className="pt-8 flex">
+            <div className="flex justify-between w-full">
+                <div className="flex">
+                    <div className="py-4">
+                        <svg className={searching ? 'animate-spin' : ''} width="16" height="16" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (gtsiolis) [inline components/dashboard/src/admin/UserSearch.tsx:70]: **praise**: LOL! Nice touch. However, this will rotate the magnifier icon when the search is taking too long. 🙈

Comment (gtsiolis) [inline components/dashboard/src/admin/UserSearch.tsx:70]: **suggestion**: A spinner before the button label to indicate a search in underway could be better! See [Spin docs](https://tailwindcss.com/docs/animation#spin). Otherwise, let's drop the `animate-spin` here or leave it for fun. Your call! 📞

File: components/dashboard/src/admin/UserDetail.tsx [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+        ]}>
+            <p>Edit user permissions by adding or removing roles for this user.</p>
+            <div className="flex flex-col">
+                {
+                    rop.map(e => <CheckBox key={e.title} title={e.title} desc="" checked={!!e.checked} onChange={e.onClick} />)
+                }
+            </div>
+        </Modal>
+        <Modal visible={viewAccountStatement} onClose={() => setViewAccountStatement(false)} title="Edit Roles" buttons={[
+            <button className="secondary" onClick={() => setViewAccountStatement(false)}>Done</button>
```
Comment (gtsiolis) [inline components/dashboard/src/admin/UserDetail.tsx:155]: **suggestion**: Could we use the same flow we already have in _Integrations_ page? For example, _Permissions_ and _Feature Flags_ could be disabled and only become active when you alter some of the permissions above. It should be fine to ship as is for now.

Currently, the page flashing on the background also slightly degrades the UX of this page when updating roles or feature flags.

File: components/dashboard/src/admin/UserDetail.tsx [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+function Label(p: { text: string, color: string }) {
+    return <div className={`ml-3 text-sm text-${p.color}-600 truncate bg-${p.color}-100 px-1.5 py-0.5 rounded-md my-auto`}>{p.text}</div>;
+}
+
+export function Property(p: { name: string, children: string | ReactChild, action?: { label: string, onClick: () => void } }) {
+    return <div className="ml-3 flex flex-col w-4/12 truncate">
+        <div className="text-base text-gray-500 truncate">
+            {p.name}
+        </div>
+        <div className="text-lg text-gray-600 font-semibold truncate">
```
Comment (gtsiolis) [inline components/dashboard/src/admin/UserDetail.tsx:175]: **issue**: We could truncate this sooner so that it does not break the rest of the interface when the list is long.

Screenshot 2021-04-05 at 11 30 33 AM

File: components/dashboard/src/admin/UserDetail.tsx [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+                        >{user.featureFlags?.permanentWSFeatureFlags?.join(', ') || '---'}</Property>
+                        <Property name="Roles"
+                            action={{
+                                label: 'Edit Roles',
+                                onClick: () => {
+                                    setEditRoles(true);
+                                }
+                            }}
+                        >{user.rolesOrPermissions?.join(', ') || '---'}</Property>
+                        <Property name="Student">{isStudent === undefined ? '---' : (isStudent ? 'Enabled' : 'Disabled')}</Property>
```
Comment (gtsiolis) [inline components/dashboard/src/admin/UserDetail.tsx:128]: **question**: Tried enabling for a gmail address but didn't work. Expected?

File: components/dashboard/src/admin/WorkspaceDetail.tsx [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    }
+    const adminLinks = getAdminLinks(workspace);
+    const adminLink = (i: number) => <Property key={'admin-'+i} name={adminLinks[i]?.name || ''}><a className="text-blue-400 hover:text-blue-600" href={adminLinks[i]?.url}>{adminLinks[i]?.title || ''}</a></Property>;
+    return <>
+        <div className="flex">
+            <div className="flex-1">
+                <div className="flex"><h3>{workspace.workspaceId}</h3><span className="my-auto ml-3"><WorkspaceStatusIndicator instance={WorkspaceAndInstance.toInstance(workspace)} /></span></div>
+                <p>{getProject(WorkspaceAndInstance.toWorkspace(workspace))}</p>
+            </div>
+            <button className="secondary ml-3" disabled={activity} onClick={reload}>Reload Data</button>
```
Comment (gtsiolis) [inline components/dashboard/src/admin/WorkspaceDetail.tsx:52]: **question**: Do we need this button? 💭

File: components/dashboard/src/admin/WorkspaceDetail.tsx [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+            <button className="danger ml-3" disabled={activity || workspace.phase === 'stopped'} onClick={stopWorkspace}>Stop Workspaces</button>
```
Comment (gtsiolis) [inline components/dashboard/src/admin/WorkspaceDetail.tsx:53]: **question**: Shouldn't this read _Stop Workspace_?

```suggestion
            <button className="danger ml-3" disabled={activity || workspace.phase === 'stopped'} onClick={stopWorkspace}>Stop Workspace</button>
```

File: components/dashboard/src/admin/UserSearch.tsx [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+            </div>
+        </div>
+        <div className="flex flex-col space-y-2">
+            <div className="px-6 py-3 flex justify-between space-x-2 text-sm text-gray-400 border-t border-b border-gray-200">
+                <div className="w-1/12"></div>
+                <div className="w-6/12">Name</div>
+                <div className="w-5/12">Created</div>
+            </div>
+        </div>
+        {searchResult.rows.filter(u => u.identities.length > 0).map(u => <UserEntry user={u} />)}
```
Comment (gtsiolis) [inline components/dashboard/src/admin/UserSearch.tsx:86]: **thought**: I still find this empty data set on load slightly confusing but I understand this helps us avoid some performance costs. Let's leave it as is for now. 🎯

File: components/dashboard/src/admin/UserDetail.tsx [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+            </div>
+        </Modal>
+    </>;
+}
+
+function Label(p: { text: string, color: string }) {
+    return <div className={`ml-3 text-sm text-${p.color}-600 truncate bg-${p.color}-100 px-1.5 py-0.5 rounded-md my-auto`}>{p.text}</div>;
+}
+
+export function Property(p: { name: string, children: string | ReactChild, action?: { label: string, onClick: () => void } }) {
```
Comment (gtsiolis) [inline components/dashboard/src/admin/UserDetail.tsx:170]: **praise**: Nice abstraction here!

File: components/dashboard/src/admin/UserDetail.tsx [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+    return <div className="ml-3 flex flex-col w-4/12 truncate">
+        <div className="text-base text-gray-500 truncate">
+            {p.name}
+        </div>
+        <div className="text-lg text-gray-600 font-semibold truncate">
+            {p.children}
```
Comment (gtsiolis) [inline components/dashboard/src/admin/UserDetail.tsx:176]: **issue**: Latest Instance ID bring in some `overflow-scroll`. Known?

File: components/dashboard/src/admin/WorkspaceDetail.tsx (hunk 12 above)
Comment (gtsiolis) [inline components/dashboard/src/admin/WorkspaceDetail.tsx:52]: **question**: No more _Download Workspace_ button?

**Review by gtsiolis — APPROVED** [review]
Estupendo! Great to this coming to life. 🌟 

Two points to consider:
1. For some reason, I cannot start a workspace anymore. 💡 
2. As long as we have the functionality in place it should be good to ship. 🚢

File: components/dashboard/src/admin/WorkspacesSearch.tsx [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+            </div>
+        </div>
+        {searchResult.rows.map(ws => <WorkspaceEntry ws={ws} />)}
+    </>
+}
+
+function WorkspaceEntry(p: { ws: WorkspaceAndInstance }) {
+    return <Link to={'/admin/workspaces/' + p.ws.workspaceId}>
+        <div className="rounded-xl whitespace-nowrap flex space-x-2 py-6 px-6 w-full justify-between hover:bg-gray-100 focus:bg-gitpod-kumquat-light group">
+            <div className="pr-3 self-center w-1/12">
```
[the remark this answers is not in the document]
Comment (gtsiolis) [reply components/dashboard/src/admin/WorkspacesSearch.tsx:111]: It seems fixed in the preview environment. 🧀

File: components/dashboard/src/admin/WorkspacesSearch.tsx [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+                <WorkspaceStatusIndicator instance={WorkspaceAndInstance.toInstance(p.ws)}/>
+            </div>
+            <div className="flex flex-col w-4/12">
+                <div className="font-medium text-gray-800 truncate hover:text-blue-600">{p.ws.workspaceId}</div>
+                <div className="text-sm overflow-ellipsis truncate text-gray-400">{getProject(WorkspaceAndInstance.toWorkspace(p.ws))}</div>
+            </div>
+            <div className="flex w-4/12 self-center">
```
[the remark this answers is not in the document]
Comment (gtsiolis) [reply components/dashboard/src/admin/WorkspacesSearch.tsx:118]: It seems fixed in the preview environment. 🧀

File: components/dashboard/src/App.tsx [hunk 19]
```diff
@@ -135,8 +145,19 @@ const renderMenu = () => (
...
-        },
-    ]}
+        }
+    ];
+
+    if (user && user?.rolesOrPermissions?.includes('admin')) {
+        left.push({
+            title: 'Admin',
+            link: '/admin',
+            alternatives: adminMenu.flatMap(e => e.link)
```
[the remark this answers is not in the document]
Comment (svenefftinge) [reply components/dashboard/src/App.tsx:155]: Yes, you can test it by logging in with another user and then remove the `admin` role from it.

File: components/dashboard/src/admin/WorkspacesSearch.tsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (svenefftinge) [reply components/dashboard/src/admin/WorkspacesSearch.tsx:111]: Yes, I fixed it already in my workspace :-)

File: components/dashboard/src/admin/WorkspacesSearch.tsx [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+    return <>
+        <div className="pt-8 flex">
+            <div className="flex justify-between w-full">
+                <div className="flex">
+                    <div className="py-4">
+                        <svg className={searching ? 'animate-spin' : ''} width="16" height="16" fill="none" xmlns="http://www.w3.org/2000/svg">
+                            <path fillRule="evenodd" clipRule="evenodd" d="M6 2a4 4 0 100 8 4 4 0 000-8zM0 6a6 6 0 1110.89 3.477l4.817 4.816a1 1 0 01-1.414 1.414l-4.816-4.816A6 6 0 010 6z" fill="#A8A29E" />
+                        </svg>
+                    </div>
+                    <input className="border-0" type="text" placeholder="Search in context urls" onKeyDown={(ke) => ke.key === 'Enter' && search() } onChange={(v) => { setSearchTerm(v.target.value) }} />
```
[the remark this answers is not in the document]
Comment (svenefftinge) [reply components/dashboard/src/admin/WorkspacesSearch.tsx:92]: I made it explicit, because it only searches in context urls so for instance not in workspace ids.

File: components/dashboard/src/admin/UserDetail.tsx (hunk 7 above)
Comment (svenefftinge) [reply components/dashboard/src/admin/UserDetail.tsx:84]: I switched that because I felt like "Delete User" is the more severe (not revertable) action. Maybe we should not have that action at all?

File: components/dashboard/src/admin/UserSearch.tsx (hunk 8 above)
Comment (svenefftinge) [reply components/dashboard/src/admin/UserSearch.tsx:70]: My cheap & sloppy attempt to indicate search progress. 🙈

Comment (svenefftinge) [reply components/dashboard/src/admin/UserSearch.tsx:70]: We need indication that search is ongoing.

File: components/dashboard/src/admin/UserDetail.tsx (hunk 9 above)
Comment (svenefftinge) [reply components/dashboard/src/admin/UserDetail.tsx:155]: I didn't introduce a dirty state that would need to be saved. It would introduce some more complexity so I opted for this simple auto-save approach.

File: components/dashboard/src/admin/UserDetail.tsx (hunk 10 above)
Comment (svenefftinge) [reply components/dashboard/src/admin/UserDetail.tsx:175]: Hey George 👋 you look so good these days 😁

File: components/dashboard/src/admin/UserDetail.tsx (hunk 11 above)
Comment (svenefftinge) [reply components/dashboard/src/admin/UserDetail.tsx:128]: Yes, it is not enabled in dev-staging

File: components/dashboard/src/admin/WorkspaceDetail.tsx (hunk 12 above)
Comment (svenefftinge) [reply components/dashboard/src/admin/WorkspaceDetail.tsx:52]: Workspace data is not updated automatically atm. We could regularly poll, but because it is the admin interface I opted for an explicit button.

File: components/dashboard/src/admin/UserSearch.tsx (hunk 14 above)
Comment (svenefftinge) [reply components/dashboard/src/admin/UserSearch.tsx:86]: We should possibly show an empty list place holder (later).

File: components/dashboard/src/admin/UserDetail.tsx (hunk 16 above)
Comment (svenefftinge) [reply components/dashboard/src/admin/UserDetail.tsx:176]: Yes, it is on purpose so I can actually see and copy the whole ID.

File: components/dashboard/src/admin/WorkspaceDetail.tsx (hunk 12 above)
Comment (svenefftinge) [reply components/dashboard/src/admin/WorkspaceDetail.tsx:52]: It doesn't work AFAIK

File: components/dashboard/src/admin/UserDetail.tsx (hunk 7 above)
Comment (gtsiolis) [reply components/dashboard/src/admin/UserDetail.tsx:97]: We could hide less prominent actions in a more actions dropdown button as we do for Workspaces. It should be fine to ship it as is and improve later. ➿

File: components/dashboard/src/admin/UserDetail.tsx (hunk 10 above)
Comment (gtsiolis) [reply components/dashboard/src/admin/UserDetail.tsx:193]: Feels comfortable over here! 😅
